src/js/integrations/stalltracking.ts

```diff
@@ -0,0 +1,336 @@
+        );
+      }
+
+      return null;
+    }
+
+    return {
+      stall_count: {
+        value: statsOnFinish.stall_count.value - statsOnStart.stall_count.value,
+      },
+      stall_total_time: {
+        value:
+          statsOnFinish.stall_total_time.value -
+          statsOnStart.stall_total_time.value,
+      },
+      stall_longest_time: statsOnFinish.stall_longest_time,
+    };
+  }
+
+  /**
+   * Logs the finish time of the span for use in `trimEnd: true` transactions.
+   */
+  private _logSpanFinish(
+    transaction: Transaction,
+    spanEndTimestamp: number
+  ): void {
+    if (
+      Math.abs(timestampInSeconds() - spanEndTimestamp) >
+      MARGIN_OF_ERROR_SECONDS
+    ) {
+      logger.log(
+        "[StallTracking] Span end not logged due to end timestamp being outside the margin of error from now."
+      );
+
+      const previousStats = this._statsAtTimestamp.get(transaction);
+
+      if (previousStats && previousStats.timestamp < spanEndTimestamp) {
+        // We also need to delete the stat for the last span, as the transaction would be trimmed to this span not the last one.
+        this._statsAtTimestamp.delete(transaction);
+      }
+    } else {
+      this._statsAtTimestamp.set(transaction, {
+        timestamp: spanEndTimestamp,
+        stats: this._getCurrentStats(transaction),
+      });
+    }
+  }
+
+  /**
+   * Get the current stats for a transaction at a given time.
+   */
+  private _getCurrentStats(transaction: Transaction): StallMeasurements {
+    return {
+      stall_count: { value: this._stallCount },
+      stall_total_time: { value: this._totalStallTime },
+      stall_longest_time: {
+        value: this._longestStallsByTransaction.get(transaction) ?? 0,
+      },
+    };
+  }
+
+  /**
+   * Start tracking stalls
+   */
+  private _startTracking(): void {
+    if (!this.isTracking) {
+      this.isTracking = true;
+      this._lastIntervalMs = Math.floor(timestampInSeconds() * 1000);
+
+      this._iteration();
+    }
+  }
+
+  /**
+   * Stops the stall tracking interval and calls reset().
+   */
+  private _stopTracking(): void {
+    this.isTracking = false;
+
+    if (this._timeout !== null) {
+      clearTimeout(this._timeout);
+      this._timeout = null;
+    }
+
+    this._reset();
+  }
+
+  /**
+   * Clears all the collected stats
+   */
+  private _reset(): void {
+    this._stallCount = 0;
+    this._totalStallTime = 0;
+    this._lastIntervalMs = 0;
+    this._longestStallsByTransaction = new Map();
+    this._runningTransactions = new Set();
+    this._statsAtTimestamp = new Map();
+  }
+
+  /**
+   * Iteration of the stall tracking interval. Measures how long the timer strayed from its expected time of running, and how
+   * long the stall is for.
+   */
+  private _iteration(): void {
+    const now = timestampInSeconds() * 1000;
+    const totalTimeTaken = now - this._lastIntervalMs;
+    const timeoutDuration = this._acceptableBusyTime / 2;
```

Comment:
Can we leave a comment why timeoutDuration is half of the acceptable busy time? I kind of understand why, but I'm sure future readers would appreciate the context.

Reply:
Hmm the original code from Facebook actually used 1/5th the "acceptable busy time", the number is pretty arbitrary and up to how often we want to run the iterations and what we consider a stall.

I came up with a better solution is to have `minimumStallThreshold` instead of `acceptableBusyTime`. `timeoutDuration` would instead be a fixed `50ms` (or whatever we want to set it to, while keeping in mind the performance impact of having a lower number, and the lower detection sensitivity of the higher number), and a stall would now instead be logged on `totalTimeTaken  > 50ms + minimumStallThreshold`. How does that sound?

Reply:
Yeah I like the `minimumStallThreshold` idea, plz include a comment in the code talking about how we switched from `acceptableBusyTime` -> `minimumStallThreshold` so we preserve this context.